text/0086-factory-pattern.md

```diff
@@ -0,0 +1,77 @@
+# Factory Pattern
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casper-network/ceps#0086](https://github.com/casper-network/ceps/pull/86)
+
+Implementing host-side support for a factory pattern.
+
+## Motivation
+
+[motivation]: #motivation
+
+This proposal introduces a host-side factory pattern to enhance the contract deployment process.
+
+Contract developers have expressed a desire to use the factory pattern, enabling scenarios like creating multiple contracts with a custom logic. While the EE technically supports this pattern, it lacks proper tests and documentation. The existing support for this was considered an undocumented feature that relies on implementation details.
+
+The proposed solution involves implementing a proper support on the host, adding relevant tests for deploying a factory contract and creating new contracts through it, using a newly introduced support in the smart contract API.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The host-side factory pattern introduces a way for contract developers to deploy contracts with custom logic through factory contracts. These factory contracts are created using a new entry point type called `EntryPointType::Install`, which marks an entry point as a factory method. Additionally, a new variant, `EntryPointType::Normal`, serves as the default entry point type and will behave like the existing `EntryPointType::Contract` in version 1.x. The proposal also introduces `EntryPointAccess::Abstract`, marking an entry point export as existing in the bytecode but not callable. This feature allows referencing WebAssembly exports from within entry points marked as `EntryPointType::Install`. Developers can even define nested factory contracts for more complex scenarios.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+In technical detail, this proposal introduces two new entry point types: `EntryPointType::Install` and `EntryPointType::Normal`. And a new entry point access type `EntryPointAccess::Abstract`.
+
+* `EntryPointType::Install` marks an entry point as a factory method responsible for deploying contracts.
+* `EntryPointType::Normal` behaves similarly to `EntryPointType::Contract` in version 1.x, while also serving as the default entry point type in version 2.0.
```

Comment:
Are we replacing the `Contract` variant with `Normal`, I think we should be specific here in saying what the go-forward will be.

Reply:
Done